File: src/main/webapp/scripts/resources/Scene.js
app.factory('Scene', function() {
  return function(data) {
    data = new DataView(data);
    var pos = 0;
    var nameLength = data.getUint8(pos++);
    this.name = new TextDecoder().decode(data.slice(pos, pos += nameLength));
    this.east = data.getUint16(pos);
    pos += 2;
    this.south = data.getUint16(pos);
    pos += 2;
    var total = this.east * this.south;
    this.tiles = [];
    var i;
    for(i = 0; i < total; i++) {
      this.tiles.push(data.getUint32(pos));
      pos += 4;
    }
    this.objects = [];
    for(i = 0; i < total; i++) {
      this.objects.push(data.getUint32(pos));
      pos += 4;
    }
  };
});

File: src/main/java/com/beijunyi/sw/output/models/Scene.java
package com.beijunyi.sw.output.models;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.KryoSerializable;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;

public class Scene implements KryoSerializable {

  private String name;
  private short east;
  private short south;
  private int[] tiles;
  private int[] objects;

  @Override
  public void write(Kryo kryo, Output output) {
    byte[] nameBytes = name.getBytes();
    int nameLength = nameBytes.length;
    if(nameLength > Byte.MAX_VALUE)
      throw new IllegalArgumentException();
    output.writeByte(nameLength);
    output.writeBytes(nameBytes);
    output.writeShort(east);
    output.writeShort(south);
    output.writeInts(tiles);
    output.writeInts(objects);
  }

  @Override
  public void read(Kryo kryo, Input input) {
    int nameLength = input.readByte();
    byte[] nameBytes = new byte[nameLength];
    input.readBytes(nameBytes);
    name = new String(nameBytes);
    east = input.readShort();
    south = input.readShort();
    int total = east * south;
    tiles = input.readInts(total);
    objects = input.readInts(total);
  }
}
